kv/kvserver: TestSnapshotAfterTruncationWithUncommittedTail failed

kv/kvserver.TestSnapshotAfterTruncationWithUncommittedTail [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9021047?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9021047?buildTab=artifacts#/) on master @ [95684388bfbb372ac8c7e1a8ebd4a0d6f447e147](https://github.com/cockroachdb/cockroach/commits/95684388bfbb372ac8c7e1a8ebd4a0d6f447e147):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3f54197]

```
Stack: 

```
goroutine 37281265 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x3?, {0x7405278, 0xc012b35c50})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x55274a0, 0xa421860})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/server.(*refreshInstanceSessionListener).OnSessionDeleted.func1({0xc009955400, 0x0})
	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:472 +0x117
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestSnapshotAfterTruncationWithUncommittedTail
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestSnapshotAfterTruncationWithUncommittedTail2652869066
    test_log_scope.go:79: use -show-logs to present logs inline
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | runtime.panicmem
*   | 	GOROOT/src/runtime/panic.go:260
*   | runtime.sigpanic
*   | 	GOROOT/src/runtime/signal_unix.go:835
*   | github.com/cockroachdb/cockroach/pkg/server.(*refreshInstanceSessionListener).OnSessionDeleted.func1
*   | 	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:472
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_amd64.s:1594
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSnapshotAfterTruncationWithUncommittedTail.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25276

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3f54197]

goroutine 37281265 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x3?, {0x7405278, 0xc012b35c50})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:229 +0x6a
panic({0x55274a0, 0xa421860})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/server.(*refreshInstanceSessionListener).OnSessionDeleted.func1({0xc009955400, 0x0})
	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:472 +0x117
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```

This code was added by @dt back in #95355, reassigning.

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -120,6 +120,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/netutil"
 	"github.com/cockroachdb/cockroach/pkg/util/netutil/addr"
 	"github.com/cockroachdb/cockroach/pkg/util/rangedesc"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -462,16 +463,28 @@ var _ slinstance.SessionEventListener = &stopperSessionEventListener{}
 func (r *refreshInstanceSessionListener) OnSessionDeleted(
 	ctx context.Context,
 ) (createAnotherSession bool) {
-	if err := r.cfg.stopper.RunAsyncTask(ctx, "refresh-instance-session", func(context.Context) {
-		nodeID, _ := r.cfg.nodeIDContainer.OptionalNodeID()
-		s, err := r.cfg.sqlLivenessProvider.Session(ctx)
-		if err != nil {
-			log.Errorf(ctx, "faild to get new liveness session ID: %v", err)
-		}
-		if _, err := r.cfg.sqlInstanceStorage.CreateNodeInstance(
-			ctx, s.ID(), s.Expiration(), r.cfg.AdvertiseAddr, r.cfg.SQLAdvertiseAddr, r.cfg.Locality, nodeID,
-		); err != nil {
-			log.Errorf(ctx, "failed to update instance with new session ID: %v", err)
+	if err := r.cfg.stopper.RunAsyncTask(ctx, "refresh-instance-session", func(ctx context.Context) {
+		for i := retry.StartWithCtx(ctx, retry.Options{MaxBackoff: time.Second * 5}); i.Next(); {
+			select {
+			case <-r.cfg.stopper.ShouldQuiesce():
+				return
+			case <-ctx.Done():
+				return
+			default:
+			}
+			nodeID, _ := r.cfg.nodeIDContainer.OptionalNodeID()
+			s, err := r.cfg.sqlLivenessProvider.Session(ctx)
+			if err != nil {
+				log.Warningf(ctx, "failed to get new liveness session ID: %v", err)
+				continue
+			}
+			if _, err := r.cfg.sqlInstanceStorage.CreateNodeInstance(
+				ctx, s.ID(), s.Expiration(), r.cfg.AdvertiseAddr, r.cfg.SQLAdvertiseAddr, r.cfg.Locality, nodeID,
+			); err != nil {
+				log.Warningf(ctx, "failed to update instance with new session ID: %v", err)
+				continue
+			}
+			return
 		}
 	}); err != nil {
 		log.Errorf(ctx, "failed to run update of instance with new session ID: %v", err)

